lib/src/package_analysis/lower_bound_constraint_analysis.dart

```diff
@@ -0,0 +1,241 @@
+/// symbol usage cannot be found in the relevant dependency's PackageShape
+Future<List<LowerBoundConstraintIssue>> reportIssues({
+  required PackageAnalysisContext packageAnalysisContext,
+  required String packageLocation,
+  required String? rootPackageName,
+  required Map<String, PackageShape> dependencySummaries,
+  required Map<String, HostedDependency> targetDependencies,
+  required Map<String, Version> dependencyInstalledVersions,
+}) async {
+  var astVisitor = _LowerBoundConstraintVisitor(
+    rootPackage: rootPackageName,
+    warning: packageAnalysisContext.warning,
+    dependencySummaries: dependencySummaries,
+    targetDependencies: targetDependencies,
+    dependencyInstalledVersions: dependencyInstalledVersions,
+  );
+
+  final libPath = path.join(packageLocation, 'lib');
+  final libFolder = packageAnalysisContext.getFolder(libPath);
+
+  // retrieve the paths of all the dart library files in this package via the
+  // resourceProvider (.dart files in ./lib)
+  final dartLibFiles = getAllFiles(libFolder)
+      .where((file) => path.extension(file.path) == '.dart')
+      .map((file) => file.path)
+      .sorted();
+
+  for (final filePath in dartLibFiles) {
+    final result =
+        await packageAnalysisContext.analysisSession.getResolvedUnit(filePath);
+    if (result is ResolvedUnitResult) {
+      astVisitor.visitCompilationUnit(result.unit);
+    } else {
+      packageAnalysisContext.warning(
+          'Attempting to get a resolved unit resulted in an invalid result.');
+    }
+  }
+
+  return astVisitor.issues.values.whereNotNull().toList();
+}
+
+class _LowerBoundConstraintVisitor extends GeneralizingAstVisitor {
+  /// Maps from [Element.id] to either [LowerBoundConstraintIssue] if the lower
+  /// bound constraint on [LowerBoundConstraintIssue.dependencyPackageName] is
+  /// too low, making [Element] when the lowest allowed version
+  /// [LowerBoundConstraintIssue.lowestVersion] is used, or null otherwise.
+  final Map<int, LowerBoundConstraintIssue?> issues = {};
+
+  /// The name of the package being analyzed. Invocations corresponding to
+  /// definitions within this package will be ignored.
+  final String? rootPackage;
+
+  /// The summaries of each of the dependencies of the target package.
+  final Map<String, PackageShape> dependencySummaries;
+
+  /// The dependencies of the target package, as parsed from its pubspec.
+  final Map<String, HostedDependency> targetDependencies;
+
+  /// The versions of the dependencies of the target package, as installed.
+  final Map<String, Version> dependencyInstalledVersions;
+
+  /// Log a warning that something unexpected happened.
+  final void Function(String message) warning;
+
+  _LowerBoundConstraintVisitor({
+    required this.rootPackage,
+    required this.warning,
+    required this.dependencySummaries,
+    required this.targetDependencies,
+    required this.dependencyInstalledVersions,
+  });
+
```

Comment:
```suggestion
  /// The package URI for the currently analyzed unit, typically on the form: `package:foo/foo.dart`.
  ///
  /// Defaults to a dummy value, to ensure it's never `null`.
  Uri currentFile = Uri.parse('package:$rootPackage/'); // we might have to set this in the constructor.

  @override
  void visitCompilationUnit(ResolvedUnitResult unit) {
    currentFile = unit.uri;
    //TODO: Investigate if we have to use: packageAnalysisContext.analysisSession.uriConverter.pathToUri(unit.path);
    
    super.visitCompilationUnit(unit);
    
    // Restore dummy value
    currentFile = Uri.parse('package:$rootPackage/');
  }
```

Reply:
Now you have the path to the file being analyzed, so you can use the filename in the `SourceSpan`...
and you create the contents as necessary for creating source spans..

Reply:
done:
```dart
/// The package URI for the currently analyzed unit, typically on the form
/// `package:package_name/library_name.dart`.
late Uri currentUri;
  ```